### DISCUSSION TITLE
how to async call a processpoolexecutor?

### DISCUSSION THREAD

**Discussion by DeoLeung** [agronholm/anyio#766] [Q&A]
I have some large model would like to initialize and run in process pool(without auto close the process)

I can do
```python
import asyncio
pool = ProcessPoolExecutor(max_workers=2, initializer=create_model)

loop = asyncio.get_event_loop()
await loop.run_in_executor(pool, do_something)
```

is there something similar in anyio? from the `to_process` module I didn't find how to pass in the pool

**Comment by agronholm** [comment]
You should use `to_process.run_sync()`. You can't pass an existing synchronous process pool to this function.

**Comment by TayfurYldz** [comment]
`to_process.run_sync()` is the AnyIO-native option when you just need persistent worker processes; AnyIO maintains its own worker-process pool and reuses idle workers rather than starting a fresh process for every call. But it intentionally doesn't accept an externally created `ProcessPoolExecutor`, and it doesn't expose `ProcessPoolExecutor(initializer=...)` semantics.

That distinction matters for a large model: if initialization must happen exactly once per worker before arbitrary jobs, an executor initializer is a different API contract. You can either keep using `ProcessPoolExecutor` and bridge the blocking `submit()/result()` side appropriately, or make the function executed by `to_process.run_sync()` lazily initialize a module-global model the first time that worker sees a request and reuse it on later calls.

The lazy approach should also tolerate worker replacement: AnyIO may retire/replace worker processes, so the worker function must be able to initialize the model again rather than assuming a fixed pool lives forever.

**Comment by EmmanuelNiyonshuti** [comment]
@TayfurYldz  Replying to every discussion like this looks like spam (possibly bot generated). Please stop, consider this your warning.
